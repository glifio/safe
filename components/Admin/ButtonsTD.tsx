import styled from 'styled-components'

export const ButtonsTD = styled.td`
  color: var(--gray-medium);
  text-align: right;

  > div {
    display: flex;
    align-items: center;
    justify-content: end;
    gap: var(--space-s);

    > span {
      cursor: pointer;
      transition: transform 0.1s ease-out;

      &:hover {
        transform: scale(1.2);
      }

      svg {
        vertical-align: middle;
      }
    }
  }
`