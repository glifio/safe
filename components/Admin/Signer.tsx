import styled from 'styled-components'
import PropTypes from 'prop-types'
import {
  IconEdit,
  IconClose,
  AddressLink,
  Address,
  GRAPHQL_ADDRESS_PROP_TYPE
} from '@glif/react-components'

const SignerEl = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  padding: 0.75em 1em 0.75em 0.75em;
  color: var(--gray-dark);
  background-color: var(--blue-gray);
  border-radius: 4px;

  div:last-child {
    display: flex;
    align-items: center;

    > span {
      display: block;
      cursor: pointer;
      line-height: 0;
      margin-left: 0.5em;
      transition: transform 0.1s ease-out;

      &:hover {
        transform: scale(1.2);
      }
    }
  }
`

export const Signer = ({ address, onRemove, onChange }: SignerProps) => (
  <SignerEl>
    <AddressLink
      id={address.id}
      address={address.robust}
      hideCopyText={false}
      hideCopy={false}
    />
    <div>
      {onChange && (
        <span role='button' aria-label='edit-signer' onClick={onChange}>
          <IconEdit stroke='var(--gray-dark)' />
        </span>
      )}
      {onRemove && (
        <span role='button' aria-label='remove-signer' onClick={onRemove}>
          <IconClose fill='var(--gray-dark)' />
        </span>
      )}
    </div>
  </SignerEl>
)

interface SignerProps {
  address: Address
  onRemove?: () => void
  onChange?: () => void
}

Signer.propTypes = {
  address: GRAPHQL_ADDRESS_PROP_TYPE.isRequired,
  onRemove: PropTypes.func,
  onChange: PropTypes.func
}
