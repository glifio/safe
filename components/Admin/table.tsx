import {
  Address,
  AddressLink,
  GRAPHQL_ADDRESS_PROP_TYPE,
  IconClose,
  IconEdit,
  isAddrEqual,
  StatusIcon,
  Wallet,
  WALLET_PROPTYPE
} from '@glif/react-components'
import styled from 'styled-components'
import PropTypes from 'prop-types'
import { SIGNER_PROPTYPE } from '../../MsigProvider/types'

const IconsWrapper = styled.div`
  display: flex;
  align-items: center;

  > span {
    display: block;
    cursor: pointer;
    line-height: 0;
    margin-left: 0.5em;
    transition: transform 0.1s ease-out;
    &:hover {
      transform: scale(1.2);
    }
  }
`

const SignerRow = ({
  signer,
  userIsSigner,
  onChange,
  onRemove
}: SignerRowProps) => {
  return (
    <tr>
      <td>
        {userIsSigner && <StatusIcon color='purple' margin='4px 0 0 1em' />}
      </td>
      <td>
        <AddressLink
          address={signer.robust || signer.id}
          shouldTruncate={false}
          hideCopyText={false}
        />
      </td>
      <td>
        <IconsWrapper>
          <span
            role='button'
            aria-label='edit-signer'
            onClick={() => onChange(signer.robust || signer.id)}
          >
            <IconEdit stroke='var(--gray-dark)' />
          </span>
          <span
            role='button'
            aria-label='remove-signer'
            onClick={() => onRemove(signer.robust || signer.id)}
          >
            <IconClose fill='var(--gray-dark)' />
          </span>
        </IconsWrapper>
      </td>
    </tr>
  )
}

type SignerRowProps = {
  signer: Address
  userIsSigner: boolean
  onRemove: (address: string) => void
  onChange: (address: string) => void
}

SignerRow.propTypes = {
  signer: GRAPHQL_ADDRESS_PROP_TYPE,
  userIsSigner: PropTypes.bool.isRequired,
  onRemove: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired
}

export const SignersTable = ({
  signers,
  wallet,
  onChange,
  onRemove
}: SignersTableProps) => {
  return (
    <table>
      <thead>
        <tr>
          <th></th>
          <th>Address</th>
          <th></th>
        </tr>
      </thead>
      <tbody>
        {signers.map((s) => (
          <SignerRow
            key={s.robust}
            signer={s}
            userIsSigner={isAddrEqual(s, wallet)}
            onChange={onChange}
            onRemove={onRemove}
          />
        ))}
      </tbody>
    </table>
  )
}

type SignersTableProps = {
  signers: Address[]
  wallet: Wallet
  onRemove: (address: string) => void
  onChange: (address: string) => void
}

SignersTable.propTypes = {
  signers: SIGNER_PROPTYPE.isRequired,
  wallet: WALLET_PROPTYPE.isRequired,
  onRemove: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired
}
