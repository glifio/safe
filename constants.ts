/* WALLET TYPES */
export const HD_WALLET = 'HD_WALLET'
export const SINGLE_KEY = 'SINGLE_KEY'

/* LOGIN OPTIONS */
export const IMPORT_MNEMONIC = 'IMPORT_MNEMONIC'
export const CREATE_MNEMONIC = 'CREATE_MNEMONIC'
export const IMPORT_SINGLE_KEY = 'IMPORT_SINGLE_KEY'

export const LEDGER = 'LEDGER'
export const METAMASK = 'METAMASK'

/* API ENDPOINTS */
export const FILSCAN = 'https://api.filscan.io:8700/v0/filscan'
export const FILSCAN_JSONRPC = 'https://api.filscan.io:8700/rpc/v1'
export const FILSCOUT = 'https://filscoutv3api.ipfsunion.cn'
export const FILFOX = 'https://filfox.info/api'

/* NETWORK VARS */
export const MAINNET = 'f'
export const TESTNET = 't'

export const GLIF_DISCORD = 'https://discord.gg/B9ju5Eu4Rq'
export const GLIF_TWITTER = 'https://twitter.com/glifio'

export const MAINNET_PATH_CODE = 461
export const TESTNET_PATH_CODE = 1

export const EXEC_ACTOR = 'f01'

/* FILECOIN APP VERSION MIN */
export const LEDGER_VERSION_MAJOR = 0
export const LEDGER_VERSION_MINOR = 18
export const LEDGER_VERSION_PATCH = 2

/* QUERY PARAMS */
export enum QPARAM {
  MSIG_ADDRESS = 'safe-id'
}

/* PAGES */
/* eslint-disable no-unused-vars */
export enum PAGE {
  SPEED_UP = '/speed-up',
  MSIG_HOME = '/home',
  MSIG_HISTORY = '/history',
  MSIG_PROPOSALS = '/proposals',
  MSIG_PROPOSAL = '/proposal',
  MSIG_ADMIN = '/admin',
  MSIG_WITHDRAW = '/withdraw',
  MSIG_APPROVE = '/approve',
  MSIG_CANCEL = '/cancel',
  MSIG_CHANGE_SIGNER = '/change-signer',
  MSIG_REMOVE_SIGNER = '/remove-signer',
  MSIG_ADD_SIGNER = '/add-signer',
  MSIG_CHANGE_APPROVALS = '/change-approvals',
  MSIG_CREATE_CONFIRM = '/create/confirm',
  MSIG_CREATE = '/create',
  MSIG_CHOOSE = '/choose',
  MSIG_CHOOSE_ACCOUNTS = '/accounts',
  CONNECT_LEDGER = '/connect/ledger',
  CONNECT_METAMASK = '/connect/metamask',
  NODE_DISCONNECTED = '/error/node-disconnected',
  LANDING = '/'
}
/* eslint-enable */
