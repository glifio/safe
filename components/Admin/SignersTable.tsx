import {
  Address,
  AddressLink,
  GRAPHQL_ADDRESS_PROP_TYPE,
  IconClose,
  IconEdit,
  isAddrEqual,
  Wallet,
  WALLET_PROPTYPE
} from '@glif/react-components'
import PropTypes from 'prop-types'
import { ButtonsTD } from './ButtonsTD'

const SignerRow = ({
  signer,
  isCurrentUser,
  onChange,
  onRemove
}: SignerRowProps) => {
  return (
    <tr>
      <td>
        <AddressLink
          address={signer.robust || signer.id}
          shouldTruncate={false}
          hideCopyText={false}
        />
      </td>
      <ButtonsTD>
        {isCurrentUser ? (
          <>this is you</>
        ) : (
          <div>
            <span
              role='button'
              aria-label='edit-signer'
              onClick={() => onChange(signer.robust || signer.id)}
            >
              <IconEdit stroke='var(--gray-dark)' />
            </span>
            <span
              role='button'
              aria-label='remove-signer'
              onClick={() => onRemove(signer.robust || signer.id)}
            >
              <IconClose fill='var(--gray-dark)' />
            </span>
          </div>
        )}
      </ButtonsTD>
    </tr>
  )
}

type SignerRowProps = {
  signer: Address
  isCurrentUser: boolean
  onRemove: (address: string) => void
  onChange: (address: string) => void
}

SignerRow.propTypes = {
  signer: GRAPHQL_ADDRESS_PROP_TYPE,
  isCurrentUser: PropTypes.bool.isRequired,
  onRemove: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired
}

export const SignersTable = ({
  signers,
  wallet,
  onChange,
  onRemove
}: SignersTableProps) => {
  return (
    <table>
      <thead>
        <tr>
          <th>Current Signers</th>
          <th></th>
        </tr>
      </thead>
      <tbody>
        {signers.map((s) => (
          <SignerRow
            key={s.robust}
            signer={s}
            isCurrentUser={isAddrEqual(s, wallet)}
            onChange={onChange}
            onRemove={onRemove}
          />
        ))}
      </tbody>
    </table>
  )
}

type SignersTableProps = {
  signers: Address[]
  wallet: Wallet
  onRemove: (address: string) => void
  onChange: (address: string) => void
}

SignersTable.propTypes = {
  signers: PropTypes.arrayOf(GRAPHQL_ADDRESS_PROP_TYPE).isRequired,
  wallet: WALLET_PROPTYPE.isRequired,
  onRemove: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired
}
