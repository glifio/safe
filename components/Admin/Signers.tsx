import PropTypes from 'prop-types'
import { useRouter } from 'next/router'
import {
  navigate,
  ADDRESS_PROPTYPE,
  Address as AddressType
} from '@glif/react-components'
import { Signer } from './Signer'
import { PAGE } from '../../constants'

export const Signers = ({ signers }: { signers: AddressType[] }) => {
  const router = useRouter()
  return (
    <>
      {signers.map((signer) => (
        <Signer
          key={signer.robust || signer.id}
          address={signer}
          onRemoveSigner={() => {
            navigate(router, {
              pageUrl: PAGE.MSIG_REMOVE_SIGNER,
              params: {
                address: signer.robust || signer.id
              }
            })
          }}
          onChangeSigner={() => {
            navigate(router, {
              pageUrl: PAGE.MSIG_CHANGE_SIGNER,
              params: {
                address: signer.robust || signer.id
              }
            })
          }}
        />
      ))}
    </>
  )
}

Signers.propTypes = {
  signers: PropTypes.arrayOf(
    PropTypes.shape({
      robust: ADDRESS_PROPTYPE,
      id: ADDRESS_PROPTYPE
    })
  ).isRequired
}
