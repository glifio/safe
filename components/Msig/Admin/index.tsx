import styled from 'styled-components'
import { useCallback, useState } from 'react'
import { ButtonV2 } from '@glif/react-components'
import {
  useWallet,
  useWalletProvider,
  reportLedgerConfigError
} from '@glif/wallet-provider-react'
import { useRouter } from 'next/router'

import { PAGE } from '../../../constants'
import { Address, Signers } from '../Shared'
import { useMsig } from '../../../MsigProvider'

const Wrapper = styled.div`
  max-width: 50rem;
  margin: 0 auto;
`

const TitleRow = styled.div`
  display: flex;
  align-items: center;
  justify-content: space-between;
  margin-top: 3em;
  h3 {
    margin: 0;
  }
`

const Title = styled.h2`
  color: var(--purple-medium);
`

const Info = styled.p`
  color: var(--gray-medium);
  margin-top: 0.25em;
`

export default function Owners() {
  const router = useRouter()
  const { NumApprovalsThreshold, Signers: signers } = useMsig()
  const wallet = useWallet()
  const { ledger, connectLedger, loginOption } = useWalletProvider()
  const [ledgerBusy, setLedgerBusy] = useState(false)
  const onShowOnLedger = useCallback(async () => {
    setLedgerBusy(true)
    const provider = await connectLedger()
    if (provider) await provider.wallet.showAddressAndPubKey(wallet.path)
    setLedgerBusy(false)
  }, [setLedgerBusy, connectLedger, wallet.path])

  const ledgerErr = reportLedgerConfigError({
    ...ledger
  })

  return (
    <div>
      <Title>Safe Admin</Title>
      <hr />

      <Wrapper>
        <TitleRow>
          <h3>Required Approvals: {NumApprovalsThreshold}</h3>
          <ButtonV2
            onClick={() => router.push(PAGE.MSIG_CHANGE_APPROVAL_THRESHOLD)}
          >
            Edit
          </ButtonV2>
        </TitleRow>
        <Info>
          The number of approvals required for a Safe proposal to execute.
        </Info>

        <TitleRow>
          <h3>Your Address</h3>
          {loginOption === 'LEDGER' && (
            <ButtonV2
              disabled={!!ledgerErr || ledgerBusy}
              onClick={onShowOnLedger}
            >
              {ledgerErr
                ? 'Ledger Device Error'
                : ledgerBusy
                ? 'Check Ledger Device'
                : 'View on Device'}
            </ButtonV2>
          )}
        </TitleRow>
        <Address address={wallet.address} />

        <TitleRow>
          <h3>Additional Signers ({signers.length - 1})</h3>
          <ButtonV2 onClick={() => router.push(PAGE.MSIG_ADD_SIGNER)}>
            Add Signer
          </ButtonV2>
        </TitleRow>
        <Info>
          These are the Filecoin addresses that can approve and reject proposals
          from your Safe.
        </Info>
        <Signers signers={signers} walletAddress={wallet.address} />
      </Wrapper>
    </div>
  )
}
