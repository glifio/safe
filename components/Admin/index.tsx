import styled from 'styled-components'
import { useRouter } from 'next/router'
import {
  ButtonV2Link,
  PageTitle,
  useWallet,
  navigate,
  ButtonRowCenter,
  WideDialog,
  LoadingScreen
} from '@glif/react-components'

import { PAGE } from '../../constants'
import { useMsig } from '../../MsigProvider'
import { SignersTable } from './SignersTable'

const RequiredApprovals = styled.div`
  margin-top: var(--space-l);

  div {
    display: flex;
    align-items: center;
    justify-content: space-between;
  }

  p {
    color: var(--gray-medium);
    margin-top: 0;
    padding-left: 1em;
  }
`

export default function Owners() {
  const { NumApprovalsThreshold, Signers: signers } = useMsig()
  const wallet = useWallet()
  const router = useRouter()

  return NumApprovalsThreshold === 0 ? (
    <LoadingScreen />
  ) : (
    <div>
      <PageTitle>Safe Admin</PageTitle>
      <hr />
      <WideDialog>
        <RequiredApprovals>
          <div>
            <h3>Required Approvals: {NumApprovalsThreshold}</h3>
            <ButtonV2Link href={PAGE.MSIG_CHANGE_APPROVALS}>Edit</ButtonV2Link>
          </div>
          <p>
            The number of approvals required for a Safe proposal to execute.
          </p>
        </RequiredApprovals>

        <div>
          <h3>Signer Addresses</h3>
          <SignersTable
            signers={signers}
            wallet={wallet}
            onRemove={(address: string) => {
              navigate(router, {
                pageUrl: PAGE.MSIG_REMOVE_SIGNER,
                params: {
                  address
                }
              })
            }}
            onChange={(address: string) => {
              navigate(router, {
                pageUrl: PAGE.MSIG_CHANGE_SIGNER,
                params: {
                  address
                }
              })
            }}
          />
          <ButtonRowCenter>
            <ButtonV2Link href={PAGE.MSIG_ADD_SIGNER}>Add signer</ButtonV2Link>
          </ButtonRowCenter>
        </div>
      </WideDialog>
    </div>
  )
}
