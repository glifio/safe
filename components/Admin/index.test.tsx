import { truncateAddress } from '@glif/react-components'
import { render, screen, act, fireEvent } from '@testing-library/react'
import {
  MULTISIG_SIGNER_ID_2,
  WALLET_ADDRESS,
  WALLET_ID_2
} from '../../test-utils/constants'
import composeMockAppTree from '../../test-utils/composeMockAppTree'
import { PAGE } from '../../constants'
import Admin from '.'

const routerPushMock = jest.fn()
jest.spyOn(require('next/router'), 'useRouter').mockImplementation(() => {
  return {
    query: {},
    pathname: PAGE.MSIG_ADMIN,
    push: routerPushMock
  }
})

describe('Admin page', () => {
  test('it renders the required approvals and the signers', () => {
    const { Tree } = composeMockAppTree('postOnboard')

    render(
      <Tree>
        <Admin />
      </Tree>
    )

    expect(screen.getByText(/Required Approvals/)).toBeInTheDocument()
    expect(screen.getByText(/Your Address/)).toBeInTheDocument()
    expect(screen.getByText(/Additional Signers/)).toBeInTheDocument()
    // signers - "t1z225tguggx4onbauimqvxzutopzdr2m4s6z6wgi" and f1nq5k2mps5umtebdovlyo7y6a3ywc7u4tobtuo3a from msig provider mocks
    // since the self signer is also listed in the top corner, it should appear twice
    expect(
      screen.getAllByText(truncateAddress(WALLET_ADDRESS)).length === 1
    ).toBeTruthy()
    expect(screen.getByText(WALLET_ID_2)).toBeInTheDocument()

    // snapshot on this test is oddly broken until https://github.com/styled-components/jest-styled-components/issues/399 is resolved
    // expect(container).toMatchSnapshot()
  })

  test('it renders a link to the add signer page', () => {
    const { Tree } = composeMockAppTree('postOnboard')

    act(() => {
      render(
        <Tree>
          <Admin />
        </Tree>
      )
      expect(screen.getByText('Add Signer').closest('a')).toHaveAttribute(
        'href',
        PAGE.MSIG_ADD_SIGNER
      )
    })
  })

  test('it renders a link to the edit req num approvals page', () => {
    const { Tree } = composeMockAppTree('postOnboard')

    act(() => {
      render(
        <Tree>
          <Admin />
        </Tree>
      )
      expect(screen.getByText('Edit').closest('a')).toHaveAttribute(
        'href',
        PAGE.MSIG_CHANGE_APPROVALS
      )
    })
  })

  test('it sends you to the change signer page with the right query params when the user clicks the edit signer button', () => {
    const { Tree } = composeMockAppTree('postOnboard')

    act(() => {
      render(
        <Tree>
          <Admin />
        </Tree>
      )
      fireEvent.click(screen.getByLabelText('edit-signer'))
    })

    expect(routerPushMock).toHaveBeenCalledWith(
      `${PAGE.MSIG_CHANGE_SIGNER}?address=${MULTISIG_SIGNER_ID_2}`
    )
  })

  test('it sends you to the remove signer page with the right query params when the user clicks the remove signer button', () => {
    const { Tree } = composeMockAppTree('postOnboard')

    act(() => {
      render(
        <Tree>
          <Admin />
        </Tree>
      )
      fireEvent.click(screen.getByLabelText('remove-signer'))
    })

    expect(routerPushMock).toHaveBeenCalledWith(
      `${PAGE.MSIG_REMOVE_SIGNER}?address=${MULTISIG_SIGNER_ID_2}`
    )
  })
})
