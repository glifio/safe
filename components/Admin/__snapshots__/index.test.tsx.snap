// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Admin page it renders the required approvals and the signers 1`] = `
.c5 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  text-align: center;
  white-space: nowrap;
  cursor: pointer;
  gap: 0.75rem;
  padding: 0.5em 0.75em;
  color: var(--black);
  border-width: 1px;
  border-style: solid;
  border-radius: 8px;
  border-color: var(--black);
  background: transparent;
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c5 svg path {
  stroke: var(--black);
}

.c5:hover {
  color: var(--purple-medium);
  border-color: var(--purple-medium);
}

.c5:hover svg path {
  stroke: var(--purple-medium);
}

.c5:active {
  color: var(--white) !important;
  border-color: var(--purple-medium) !important;
  background: var(--purple-medium) !important;
}

.c5:active svg path {
  stroke: var(--white) !important;
}

.c5:hover {
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c10 {
  width: 24;
  height: 24;
}

.c12 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: var(--space-m);
  margin: 1.5em 0;
}

.c13 {
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c7 h4 {
  margin: 0;
  color: var(--gray-dark);
}

.c7 p {
  margin: 0;
  line-height: 1.5;
}

.c9 {
  position: relative;
  cursor: pointer;
}

.c9 svg {
  -webkit-transition: -webkit-transform 0.1s ease-out;
  -webkit-transition: transform 0.1s ease-out;
  transition: transform 0.1s ease-out;
}

.c9 .copy-text {
  position: absolute;
  top: calc(50% - 0.2em);
  left: calc(100% + 0.25em);
  color: var(--purple-medium);
  opacity: 0;
  -webkit-transform: translateY(-50%);
  -ms-transform: translateY(-50%);
  transform: translateY(-50%);
  -webkit-transition: opacity 0.2s ease-in;
  transition: opacity 0.2s ease-in;
  pointer-events: none;
}

.c9:hover svg {
  -webkit-transform: scale(1.25);
  -ms-transform: scale(1.25);
  transform: scale(1.25);
}

.c9:hover .copy-text {
  opacity: 1;
  pointer-events: all;
}

.c8 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  gap: 0.25em;
  line-height: 1.5;
}

.c8 > a {
  color: var(--purple-medium);
  -webkit-text-decoration: none;
  text-decoration: none;
}

.c8 > a:hover {
  color: var(--purple-medium);
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c8 > a svg {
  -webkit-transition: 0.24s ease-in-out;
  transition: 0.24s ease-in-out;
  vertical-align: middle;
}

.c8 > a svg:hover {
  -webkit-transform: scale(1.25);
  -ms-transform: scale(1.25);
  transform: scale(1.25);
  fill: var(--purple-medium);
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: var(--space-m);
  width: 100%;
  margin: 0 auto;
  max-width: 40em;
}

.c2 form {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  gap: var(--space-m);
}

.c3 {
  max-width: 50em;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
  gap: var(--space-l);
  margin: 1em 0;
  font-size: var(--font-size-xl);
}

.c0 .side-bar {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: var(--space-m);
  font-size: 1rem;
}

.c1 {
  margin: 0;
  color: var(--purple-medium);
}

.c11 {
  color: var(--gray-medium);
  text-align: right;
}

.c11 > div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: end;
  -webkit-justify-content: end;
  -ms-flex-pack: end;
  justify-content: end;
  gap: var(--space-s);
}

.c11 > div > span {
  cursor: pointer;
  -webkit-transition: -webkit-transform 0.1s ease-out;
  -webkit-transition: transform 0.1s ease-out;
  transition: transform 0.1s ease-out;
}

.c11 > div > span:hover {
  -webkit-transform: scale(1.2);
  -ms-transform: scale(1.2);
  transform: scale(1.2);
}

.c11 > div > span svg {
  vertical-align: middle;
}

.c6 > div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  gap: var(--space-s);
}

.c4 {
  margin-top: var(--space-l);
}

.c4 div {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  -ms-flex-pack: justify;
  justify-content: space-between;
}

.c4 p {
  color: var(--gray-medium);
  margin-top: 0;
  padding-left: 1em;
}

<div>
  <div>
    <div
      class="c0"
    >
      <h2
        class="c1"
      >
        Safe Admin
      </h2>
    </div>
    <hr />
    <div
      class="c2 c3"
    >
      <div
        class="c4"
      >
        <div>
          <h3>
            Required Approvals: 
            2
          </h3>
          <a
            class="c5"
            href="/change-approvals"
          >
            Edit
          </a>
        </div>
        <p>
          The number of approvals required for a Safe proposal to execute.
        </p>
      </div>
      <div>
        <h3>
          Signer Addresses
        </h3>
        <table>
          <tbody>
            <tr>
              <td
                class="c6"
              >
                <div>
                  <span>
                    (You)
                  </span>
                  <div
                    class="c7"
                  >
                    <div
                      class="c8"
                      color="var(--purple-medium)"
                    >
                      <a
                        href="https://explorer-calibration.glif.link/actor/?address=t1z225tguggx4onbauimqvxzutopzdr2m4s6z6wgi"
                        rel="noreferrer noopener"
                        target="_blank"
                      >
                        t1z225tguggx4onbauimqvxzutopzdr2m4s6z6wgi
                      </a>
                      <span
                        class="c9"
                        color="var(--purple-medium)"
                        role="button"
                      >
                        <svg
                          class="c10"
                          fill="none"
                          height="24"
                          viewBox="0 0 24 24"
                          width="24"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <rect
                            fill="#1A0066"
                            height="12"
                            width="12"
                            x="8"
                            y="4"
                          />
                          <rect
                            fill="#E0D7FE"
                            height="12"
                            width="12"
                            x="4"
                            y="8"
                          />
                        </svg>
                        <span
                          class="copy-text"
                        >
                          Copy
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </td>
              <td
                class="c11"
              >
                <div>
                  <span
                    aria-label="edit-signer"
                    role="button"
                  >
                    <svg
                      class="c10"
                      fill="none"
                      height="24"
                      stroke="var(--gray-dark)"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M15.232 5.232l3.536 3.536m-2.036-5.036a2.5 2.5 0 113.536 3.536L6.5 21.036H3v-3.572L16.732 3.732z"
                        fill="none"
                        fill-rule="evenodd"
                        stroke-width="2"
                      />
                    </svg>
                  </span>
                  <span
                    aria-label="remove-signer"
                    role="button"
                  >
                    <svg
                      class="c10"
                      fill="var(--gray-dark)"
                      height="24"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M19.7333 4.2669C19.3776 3.91103 18.801 3.91103 18.4453 4.2669L11.9994 10.7166L5.55552 4.26885C5.19986 3.91299 4.62323 3.91299 4.26757 4.26885C3.91191 4.62472 3.91191 5.20169 4.26757 5.55756L10.7115 12.0053L4.27793 18.4426C3.92228 18.7985 3.92228 19.3755 4.27793 19.7313C4.63359 20.0872 5.21022 20.0872 5.56587 19.7313L11.9994 13.294L18.435 19.7333C18.7906 20.0892 19.3672 20.0892 19.7229 19.7333C20.0786 19.3774 20.0786 18.8005 19.7229 18.4446L13.2874 12.0053L19.7333 5.5556C20.0889 5.19974 20.0889 4.62276 19.7333 4.2669Z"
                        fill="var(--gray-dark)"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </span>
                </div>
              </td>
            </tr>
            <tr>
              <td
                class="c6"
              >
                <div>
                  <div
                    class="c7"
                  >
                    <div
                      class="c8"
                      color="var(--purple-medium)"
                    >
                      <a
                        href="https://explorer-calibration.glif.link/actor/?address=t1mbk7q6gm4rjlndfqw6f2vkfgqotres3fgicb2uq"
                        rel="noreferrer noopener"
                        target="_blank"
                      >
                        t1mbk7q6gm4rjlndfqw6f2vkfgqotres3fgicb2uq
                      </a>
                      <span
                        class="c9"
                        color="var(--purple-medium)"
                        role="button"
                      >
                        <svg
                          class="c10"
                          fill="none"
                          height="24"
                          viewBox="0 0 24 24"
                          width="24"
                          xmlns="http://www.w3.org/2000/svg"
                        >
                          <rect
                            fill="#1A0066"
                            height="12"
                            width="12"
                            x="8"
                            y="4"
                          />
                          <rect
                            fill="#E0D7FE"
                            height="12"
                            width="12"
                            x="4"
                            y="8"
                          />
                        </svg>
                        <span
                          class="copy-text"
                        >
                          Copy
                        </span>
                      </span>
                    </div>
                  </div>
                </div>
              </td>
              <td
                class="c11"
              >
                <div>
                  <span
                    aria-label="edit-signer"
                    role="button"
                  >
                    <svg
                      class="c10"
                      fill="none"
                      height="24"
                      stroke="var(--gray-dark)"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M15.232 5.232l3.536 3.536m-2.036-5.036a2.5 2.5 0 113.536 3.536L6.5 21.036H3v-3.572L16.732 3.732z"
                        fill="none"
                        fill-rule="evenodd"
                        stroke-width="2"
                      />
                    </svg>
                  </span>
                  <span
                    aria-label="remove-signer"
                    role="button"
                  >
                    <svg
                      class="c10"
                      fill="var(--gray-dark)"
                      height="24"
                      viewBox="0 0 24 24"
                      width="24"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M19.7333 4.2669C19.3776 3.91103 18.801 3.91103 18.4453 4.2669L11.9994 10.7166L5.55552 4.26885C5.19986 3.91299 4.62323 3.91299 4.26757 4.26885C3.91191 4.62472 3.91191 5.20169 4.26757 5.55756L10.7115 12.0053L4.27793 18.4426C3.92228 18.7985 3.92228 19.3755 4.27793 19.7313C4.63359 20.0872 5.21022 20.0872 5.56587 19.7313L11.9994 13.294L18.435 19.7333C18.7906 20.0892 19.3672 20.0892 19.7229 19.7333C20.0786 19.3774 20.0786 18.8005 19.7229 18.4446L13.2874 12.0053L19.7333 5.5556C20.0889 5.19974 20.0889 4.62276 19.7333 4.2669Z"
                        fill="var(--gray-dark)"
                        fill-rule="evenodd"
                      />
                    </svg>
                  </span>
                </div>
              </td>
            </tr>
          </tbody>
        </table>
        <div
          class="c12 c13"
        >
          <a
            class="c5"
            href="/add-signer"
          >
            Add signer
          </a>
        </div>
      </div>
    </div>
  </div>
</div>
`;
